server/src/controllers/attainable.ts

```diff
@@ -151,6 +161,70 @@ export async function addAttainable(req: Request, res: Response): Promise<void>
   });
 }
 
+export async function deleteAttainable(req: Request, res: Response): Promise<void> {
+  /*
+   * TODO: Check that the requester is logged in, 401 Unauthorized if not
+   * TODO: Check that the requester is authorized to add a course instance, 403
+   * Forbidden if not
+   */
+
+  // Get path parameters.
+  const courseId: number = Number(req.params.courseId);
+  const courseInstanceId: number = Number(req.params.instanceId);
+  const attainableId: number = Number(req.params.attainmentId);
+
+  // Validation.
+  await idSchema.validate({ id: courseId }, { abortEarly: false });
+  await idSchema.validate({ id: courseInstanceId }, { abortEarly: false });
+  await idSchema.validate({ id: attainableId }, { abortEarly: false });
+  await validateCourseAndInstance(courseId, courseInstanceId);
+
+  const attainable: Attainable = await findAttainableById(attainableId, HttpCode.NotFound);
+
+  // IDs of attainables to be deleted.
+  const ids: Array<number> = [attainable.id];
+
+  function findAllSubAttainables(
+    parentId: number,
+    allAttainables: Array<Attainable>
+  ): void {
+    for (const otherAttainable of allAttainables) {
+      if (otherAttainable.attainableId === parentId) {
+        ids.push(otherAttainable.id);
+        findAllSubAttainables(otherAttainable.id, allAttainables);
+      }
+    }
+  }
+
+  findAllSubAttainables(
+    attainable.id,
+    await Attainable.findAll({
+      attributes: ['id', 'attainable_id'],
+      where: {
+        courseInstanceId: courseInstanceId
+      }
+    })
+  );
+
+  /*
+   * TODO: Also delete grades of these attainables.
+   * TODO: Query inside for loop may be bad, is there a better way to delete
+   * multiple entries?
```

Comment:
Since we have set the sequelize association for attainable parents as:
```javascript
Attainable.hasMany(Attainable, {
  onDelete: 'CASCADE',
  onUpdate: 'CASCADE'
});

Attainable.belongsTo(Attainable, {
  targetKey: 'id',
  foreignKey: 'attainableId'
});
```

The following should be enough for deleting the attainable and all of its subAttainables.
```javascript
await attainable.destroy();
```

Reply:
Seems to work! Will this also delete the grades referencing these attainments ("attainables")?

Reply:
Yes, that associations is also set to cascade on delete 👍